velox/functions/prestosql/aggregates/GeometricMeanAggregate.cpp

```diff
@@ -0,0 +1,144 @@
+
+  static bool toIntermediate(
+      exec::out_type<Row<double, int64_t>>& out,
+      exec::arg_type<T> in) {
+    out.copy_from(std::make_tuple(static_cast<double>(in), 1));
+    return true;
+  }
+
+  struct AccumulatorType {
+    // Sum of `log(value)`.
+    double logSum_;
+    // Count of values.
+    int64_t count_;
+
+    AccumulatorType() = delete;
+
+    explicit AccumulatorType(HashStringAllocator* /*allocator*/) {
+      logSum_ = 0;
+      count_ = 0;
+    }
+
+    void addInput(HashStringAllocator* /*allocator*/, exec::arg_type<T> data) {
+      logSum_ += std::log(data);
+      count_ = checkedPlus<int64_t>(count_, 1);
+    }
+
+    void combine(
+        HashStringAllocator* /*allocator*/,
+        exec::arg_type<Row<double, int64_t>> other) {
+      VELOX_CHECK(other.at<0>().has_value());
+      VELOX_CHECK(other.at<1>().has_value());
+      logSum_ += other.at<0>().value();
+      count_ = checkedPlus<int64_t>(count_, other.at<1>().value());
+    }
+
+    bool writeFinalResult(exec::out_type<OutputType>& out) {
+      out = std::exp(logSum_ / count_);
+      return true;
+    }
+
+    bool writeIntermediateResult(exec::out_type<IntermediateType>& out) {
+      out = std::make_tuple(logSum_, count_);
+      return true;
+    }
+  };
+};
+
+} // namespace
+
+void registerGeometricMeanAggregate(const std::string& prefix) {
+  const std::string name = prefix + kGeometricMean;
+
+  std::vector<std::shared_ptr<exec::AggregateFunctionSignature>> signatures;
+
+  for (const auto& inputType : {"bigint", "double"}) {
+    signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                             .returnType("double")
+                             .intermediateType("row(double,bigint)")
+                             .argumentType(inputType)
+                             .build());
+  }
+
+  exec::registerAggregateFunction(
+      name,
+      std::move(signatures),
+      [name](
+          core::AggregationNode::Step step,
+          const std::vector<TypePtr>& argTypes,
+          const TypePtr& resultType,
+          const core::QueryConfig& /*config*/)
+          -> std::unique_ptr<exec::Aggregate> {
+        VELOX_CHECK_EQ(argTypes.size(), 1, "{} takes one argument", name);
+        auto inputType = argTypes[0];
+        if (exec::isRawInput(step)) {
+          switch (inputType->kind()) {
+            case TypeKind::BIGINT:
+              return std::make_unique<
+                  SimpleAggregateAdapter<GeometricMeanAggregate<int64_t>>>(
+                  resultType);
+            case TypeKind::DOUBLE:
+              return std::make_unique<
+                  SimpleAggregateAdapter<GeometricMeanAggregate<double>>>(
+                  resultType);
+            default:
+              VELOX_FAIL(
+                  "Unknown input type for {} aggregation {}",
+                  name,
+                  inputType->kindName());
+          }
+        } else {
+          switch (resultType->kind()) {
```

Comment:
For intermediate agg, resultType will be ROW. Can you make sure to handle this?

Reply:
done

Reply:
done